pkg/shp/suggestion/suggest.go

```diff
@@ -0,0 +1,123 @@
+package suggestion
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// SubcommandsRequiredWithSuggestions will ensure we have a subcommand provided by the user and augments it with
+// suggestion for commands, alias and help on root command.
+func SubcommandsRequiredWithSuggestions(cmd *cobra.Command, args []string) error {
+	requireMsg := "unknown command \"%s\" for \"%s\""
+	typedName := ""
+	// This will be triggered if cobra didn't find any subcommands.
+	// Find some suggestions.
+	var suggestions []string
+
+	if len(args) != 0 && !cmd.DisableSuggestions {
+		typedName += args[0]
+		if cmd.SuggestionsMinimumDistance <= 0 {
+			cmd.SuggestionsMinimumDistance = 2
+		}
+		// subcommand suggestions
+		suggestions = cmd.SuggestionsFor(args[0])
+
+		// subcommand alias suggestions (with distance, not exact)
+		for _, c := range cmd.Commands() {
+			if c.IsAvailableCommand() {
+				candidate := suggestsByPrefixOrLd(typedName, c.Name(), cmd.SuggestionsMinimumDistance)
+				if candidate == "" {
+					continue
+				}
+				_, found := Find(suggestions, candidate)
+				if !found {
+					suggestions = append(suggestions, candidate)
+				}
+			}
+		}
+
+		// help for root command
+		if !cmd.HasParent() {
+			candidate := suggestsByPrefixOrLd(typedName, "help", cmd.SuggestionsMinimumDistance)
+			if candidate != "" {
+				suggestions = append(suggestions, candidate)
+			}
+		}
+	}
+
+	var suggestionsMsg string
+	if len(suggestions) > 0 {
+		suggestionsMsg += "\nDid you mean this?\n"
+		for _, s := range suggestions {
+			suggestionsMsg += fmt.Sprintf("\t%v\n", s)
+		}
+	}
+
+	if suggestionsMsg != "" {
+		requireMsg = fmt.Sprintf("%s\n%s", requireMsg, suggestionsMsg)
+		return fmt.Errorf(requireMsg, typedName, cmd.CommandPath())
+	}
+
+	return cmd.Help()
+}
+
+// suggestsByPrefixOrLd suggests a command by levenshtein distance or by prefix.
+// It returns an empty string if nothing was found
+func suggestsByPrefixOrLd(typedName, candidate string, minDistance int) string {
+	levenshteinVariable := levenshteinDistance(typedName, candidate, true)
+	suggestByLevenshtein := levenshteinVariable <= minDistance
+	suggestByPrefix := strings.HasPrefix(strings.ToLower(candidate), strings.ToLower(typedName))
+	if !suggestByLevenshtein && !suggestByPrefix {
+		return ""
+	}
+	return candidate
+}
+
+// ld compares two strings and returns the levenshtein distance between them.
+func levenshteinDistance(s, t string, ignoreCase bool) int {
```

Comment:
I'd prefer we use an external library for such a well defined algorithm. We can keep the ignoreCase logic by applying `strings.ToLower` before calculating the distance.

https://github.com/texttheater/golang-levenshtein appears to be a good solution.

Reply:
There is already an issue to track this https://github.com/spf13/cobra/issues/981 and a PR to fix this https://github.com/spf13/cobra/pull/841. Once this is merged then probably we can revert this in future.

Reply:
addressed :+1:

Reply:
I agree with @adambkaplan on using a well-defined algorithm, specially something that's common out there.